File: unk/unk.c
#include <stdio.h>
#include <string.h>

int main()
{
    char c;
    char pattern[] = "<unk>";
    char replacement[] = "<raw_unk>"; 
    char buf[5]; // length of "<unk>"
    int i, j, pat_len;

    pat_len = strlen(pattern);
    for (i = 0; (c = getchar()) != EOF;) {
        if (c == pattern[i]) {
            buf[i++] = c;
            if (i == pat_len) {
                printf("%s", replacement);
                i = 0;
            }
        } else {
            if (i > 0) {
                for (j = 0; j < i; j++) {
                    putchar(buf[j]);
                }
                i = 0;
            }
            putchar(c);
        }
    }

    return 0;
}

File: unk/unk.go
// unk.go is a possible solution for a specific problem posted on StackExchange
// https://unix.stackexchange.com/questions/413664/replace-string-in-a-huge-70gb-one-line-text-file
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

func main() {
	const (
		pattern     = "<unk>"
		replacement = "<raw_unk>"
	)
	var match int
	var char rune
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanRunes)
	for scanner.Scan() {
		char = rune(scanner.Text()[0])
		if char == []rune(pattern)[match] {
			match++
			if match == len(pattern) {
				fmt.Print(replacement)
				match = 0
			}
		} else {
			if match > 0 {
				fmt.Print(string(pattern[:match]))
				match = 0
			}
			if char == rune(pattern[0]) {
				match = 1
			} else {
				fmt.Print(string(char))
			}
		}
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
}
